[FLINK-39636][table] Add produces_full_deletes parameter in TO_CHANGELOG (#28235)
apache/flink: docs/content/docs/sql/reference/queries/changelog.md

| `input`                 | Yes      | The input table. With `PARTITION BY`, rows with the same key are co-located and run in the same operator instance. Without `PARTITION BY`, each row is processed independently. Accepts insert-only, retract, and upsert tables. For upsert tables, the provided `PARTITION BY` key should match or be a subset of the upsert key of the subquery.       |
| `op`                    | No       | A `DESCRIPTOR` with a single column name for the operation code column. Defaults to `op`.                                                                                                                                                                                                                                                                |
| `op_mapping`            | No       | A `MAP<STRING, STRING>` mapping change operation names to custom output codes. Keys can contain comma-separated names to map multiple operations to the same code (e.g., `'INSERT, UPDATE_AFTER'`). When provided, only mapped operations are forwarded - unmapped events are dropped. Each change operation may appear at most once across all entries. |

> | `produces_full_deletes` | No       | A `BOOLEAN` literal that controls how DELETE rows are emitted. When `true` (default), the function requires fully-populated DELETE rows from the input. The planner inserts a `ChangelogNormalize` operator for upsert sources that emit key-only deletes, so downstream sees the full pre-image on DELETE. When `false`, the function instead emits partial DELETE rows: in row semantics it preserves the planner-derived upsert key columns of the input and nulls the rest; in set semantics (`PARTITION BY`) it preserves the partition key and nulls the rest. Requires that the input declares an upsert key or that the call uses `PARTITION BY`; otherwise the function has no identifying columns to preserve and the call is rejected. |

Reviewer: Let's have here a short description with the information that the user needs to use it. Detailed information is in the section below 

```suggestion
| `produces_full_deletes` | No | A `BOOLEAN` literal that controls how DELETE rows are emitted. When `true` (default), DELETE rows carry all columns, the full image. When `false`, only the identifying key columns are preserved and the rest are nulled. See [Delete handling](#delete-handling) for more details. |

```

Author: Done

In **row semantics** (no `PARTITION BY`), the function preserves the upsert key columns of the input on DELETE rows and nulls the rest. The upsert key is the one the planner already derives from the input (typically a declared `PRIMARY KEY`), so no `PARTITION BY` is required to get partial DELETEs.

```sql
-- Source 't' declares PRIMARY KEY (id). Source emits -D[id:5, name:'Alice'] (full pre-image).
-- Output: +I[op:'DELETE', id:5, name:null]
SELECT * FROM TO_CHANGELOG(input => TABLE t)
```

When the input has no derivable upsert key (e.g. a pure append-only source, or an upstream operator that erased the key), there is no identifying column to preserve. The function then passes the input through unchanged.

```sql
-- Source emits -D[id:5] (key-only, no declared key).
-- Output: +I[op:'DELETE', id:5, name:null]
SELECT * FROM TO_CHANGELOG(input => TABLE upsert_source)

> ```

Reviewer: The example here is not matching what you want to show. You mention it passes through unchanged but the example does name:null

Author: I re-wrote the docs on deletion. Explaining 

<table>
<tr> 
 <td>  /
 <td> <b>produces_full_deletes=true</b>
 <td> <b>produces_full_deletes=false</b>
<tr>
 <td> <b>Set Semantic (PARTITION BY)</b>
 <td> Source emits Upsert/Retract + (No)Need ChangelogNormalizer + Output
 <td> Source emits Upsert/Retract + (No)Need ChangelogNormalizer + Output
<tr>
 <td> <b>Row Semantic (no PARTITION BY)</b>
 <td> Source emits Upsert/Retract + (No)Need ChangelogNormalizer + Output
 <td> Source emits Upsert/Retract + (No)Need ChangelogNormalizer + Output
</table>

-- UPDATE_BEFORE is dropped (not in the mapping)
```

> #### Delete handling

Reviewer: Nice documentation!

```suggestion
#### Full vs partial deletes
```

Author: Updated the title

**Row semantics** (no `PARTITION BY`):

```sql
-- Upsert source: -D[id:5] (key-only).
-- ChangelogNormalize materializes the full pre-image from state.
-- Output: +I[op:'DELETE', id:5, name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE upsert_source)

-- Retract source: -D[id:5, name:'Alice'] (full pre-image).
-- No ChangelogNormalize inserted; the input row is passed through unchanged.
-- Output: +I[op:'DELETE', id:5, name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE retract_source)
```

**Set semantics** (`PARTITION BY`):

```sql
-- Upsert source: -D[id:5] (key-only).
-- ChangelogNormalize materializes the full pre-image from state.
-- Output: +I[id:5, op:'DELETE', name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE upsert_source PARTITION BY id)

-- Retract source: -D[id:5, name:'Alice'] (full pre-image).
-- No ChangelogNormalize inserted; the input row is passed through unchanged.
-- Output: +I[id:5, op:'DELETE', name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE retract_source PARTITION BY id)
```

##### `produces_full_deletes => false`

> The planner skips `ChangelogNormalize` and the function emits partial DELETE rows. This avoids the stateful normalization operator for upsert sources (e.g. Kafka compacted topics) where the full pre-image is not needed downstream. Requires an upsert key (row semantics) or `PARTITION BY` (set semantics); otherwise the call is rejected with a validation error.

Reviewer: ```suggestion
The planner skips `ChangelogNormalize` and the function emits partial DELETE rows. This avoids the stateful normalization operator for upsert sources (e.g. Kafka compacted topics) where the full pre-image is not needed downstream. Requires an upsert key to be present for the input table (row semantics) or `PARTITION BY` (set semantics); otherwise the call is rejected with a validation error.
```

Reviewer: I think we don't have a good explanation of the upsert key in the docs. It'd be good to have it and then link it in the places where we mention it for the users

Author: Added a section for upsertKey

> **Row semantics** (no `PARTITION BY`): the function preserves the planner-derived upsert key columns on DELETE rows and nulls the rest. The upsert key is typically a declared `PRIMARY KEY`.

Reviewer: ```suggestion
**Row semantics** (no `PARTITION BY`): the function preserves the planner-derived upsert key columns on DELETE rows and nulls the rest. The upsert key is typically a declared `PRIMARY KEY` when directly reading from source or the key provided in a GROUP BY <key>.
```

Author: added

```sql
-- Upsert source with PRIMARY KEY (id): -D[id:5] (key-only).
-- Output: +I[op:'DELETE', id:5, name:null]
SELECT * FROM TO_CHANGELOG(input => TABLE upsert_source, produces_full_deletes => false)

-- Retract source with PRIMARY KEY (id): -D[id:5, name:'Alice'] (full pre-image).
-- Output: +I[op:'DELETE', id:5, name:null]
SELECT * FROM TO_CHANGELOG(input => TABLE retract_source, produces_full_deletes => false)
```

> **Set semantics** (`PARTITION BY`): the function preserves the partition key and nulls every non-partition-key column on DELETE rows. This matches the shape expected by upsert sinks and Kafka compacted topics.

Reviewer: ```suggestion
**Set semantics** (`PARTITION BY`): the function preserves the partition key and nulls every non-partition-key column on DELETE rows. The key used as the partition-key column should be the unique key that will be used as the record identifier. This matches the shape expected by upsert sinks and Kafka compacted topics.
```

Author: done

-- UPDATE_BEFORE is dropped (not in the mapping)
```

#### Upsert key

An **upsert key** is a column or set of columns that uniquely identifies a row across its lifecycle in a changelog. It is what downstream operators and sinks use to decide which earlier row a new INSERT, UPDATE_AFTER, or DELETE refers to.

The planner derives the upsert key from the input table:

* A declared `PRIMARY KEY` on the source table when reading directly.
* The grouping columns of an upstream `GROUP BY <key>`.

> * The keys propagated by operators that preserve them (e.g. lookup joins, calc-projections that keep the key columns).

Reviewer 2: ```suggestion
* The grouping columns of an upstream `GROUP BY <key>` or `PARTITION BY <key>`.
* The keys propagated by operators that preserve them (e.g. lookup joins, projections that keep the key columns).
```

When no upsert key can be derived (e.g. a plain append-only source with no key constraint and no grouping upstream), the input has no row identity and downstream operators must treat it as append-only or fall back to retract semantics.

> `TO_CHANGELOG` consumes the upsert key to decide which columns to preserve when emitting partial DELETE rows. See [Full vs partial deletes](#full-vs-partial-deletes) below.

Reviewer 2: I would drop this paragraph. Imho it doesn't contain any interesting information.

#### Full vs partial deletes

The `produces_full_deletes` argument controls how DELETE rows are emitted and what the planner requires from the input. The matrix below shows each combination with `PARTITION BY` (set semantics) and without (row semantics). When `false`, the function relies on the input table's [upsert key](#upsert-key) to decide which columns to preserve.

##### `produces_full_deletes => true` (default)

> The planner requires fully-populated DELETE rows on the input. For upsert sources that emit key-only deletes, a `ChangelogNormalize` operator is inserted upstream to materialize the full pre-image from state. For sources that already emit a full pre-image (e.g. retract), the flag is a no-op. The function then passes the input row through unchanged on DELETE.

Reviewer 2: ```suggestion
The planner requires fully-populated DELETE rows on the input. For upsert sources that emit key-only deletes, a `ChangelogNormalize` operator is inserted upstream to calculate the full pre-image from state. For sources that already emit a full pre-image (e.g. retract), the flag is a no-op. The function then passes the input row through unchanged on DELETE.
```

**Row semantics** (no `PARTITION BY`):

```sql

> -- Upsert source: -D[id:5] (key-only).

Reviewer 2: I think this would make it clearer.
```suggestion
-- Upsert source: -D[id:5, name:null] (key-only).
```

> -- ChangelogNormalize materializes the full pre-image from state.

Reviewer 2: ```suggestion
-- ChangelogNormalize calculates the full pre-image from state.
```

-- Output: +I[op:'DELETE', id:5, name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE upsert_source)

-- Retract source: -D[id:5, name:'Alice'] (full pre-image).
-- No ChangelogNormalize inserted; the input row is passed through unchanged.
-- Output: +I[op:'DELETE', id:5, name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE retract_source)
```

**Set semantics** (`PARTITION BY`):

```sql
-- Upsert source: -D[id:5] (key-only).
-- ChangelogNormalize materializes the full pre-image from state.
-- Output: +I[id:5, op:'DELETE', name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE upsert_source PARTITION BY id)

-- Retract source: -D[id:5, name:'Alice'] (full pre-image).
-- No ChangelogNormalize inserted; the input row is passed through unchanged.
-- Output: +I[id:5, op:'DELETE', name:'Alice']
SELECT * FROM TO_CHANGELOG(input => TABLE retract_source PARTITION BY id)
```

##### `produces_full_deletes => false`

> The planner skips `ChangelogNormalize` and the function emits partial DELETE rows. This avoids the stateful normalization operator for upsert sources (e.g. Kafka compacted topics) where the full pre-image is not needed downstream. Requires an [upsert key](#upsert-key) to be present for the input table (row semantics) or `PARTITION BY` (set semantics); otherwise the call is rejected with a validation error.

Reviewer 2: ```suggestion
The planner skips `ChangelogNormalize` and the function emits partial DELETE rows. This avoids the stateful normalization operator for upsert sources (e.g. Kafka compacted topics) where the full pre-image is not needed downstream. This requires an [upsert key](#upsert-key) to be present for the input table (row semantics) or `PARTITION BY` (set semantics); otherwise the call is rejected with a validation error.
```

**Row semantics** (no `PARTITION BY`): the function preserves the planner-derived upsert key columns on DELETE rows and nulls the rest. The upsert key is typically a declared `PRIMARY KEY` when directly reading from a source or the key provided in a `GROUP BY <key>`.

```sql
-- Upsert source with PRIMARY KEY (id): -D[id:5] (key-only).
-- Output: +I[op:'DELETE', id:5, name:null]
SELECT * FROM TO_CHANGELOG(input => TABLE upsert_source, produces_full_deletes => false)

-- Retract source with PRIMARY KEY (id): -D[id:5, name:'Alice'] (full pre-image).
-- Output: +I[op:'DELETE', id:5, name:null]
SELECT * FROM TO_CHANGELOG(input => TABLE retract_source, produces_full_deletes => false)
```

> **Set semantics** (`PARTITION BY`): the function preserves the partition key and nulls every non-partition-key column on DELETE rows. The key used as the partition-key column should be the unique key that will be used as the record identifier. This matches the shape expected by upsert sinks and Kafka compacted topics.

Reviewer 2: ```suggestion
**Set semantics** (`PARTITION BY`): the function preserves the partition key and nulls every non-partition-key column on DELETE rows. The key used as the partition-key column should be the unique key that will be used as the record identifier. This matches the shape expected by upsert sinks such as Kafka compacted topics.
```

    map("INSERT, UPDATE_AFTER", "false", "DELETE", "true").asArgument("op_mapping")
);

// Require fully-populated DELETE rows from the input (inserts a ChangelogNormalize for

> // upsert sources). When false (default), no full-delete requirement is enforced; in row

Reviewer: "When false (default)" still out of date. Can you double check we changed the default and adjusted docs everywhere?

Author: Updated all the docs and everything should be aligned now!
